**сначала надо скачать викитекст в корень**

In [1]:
import sys
sys.path.append('./task2')

In [2]:
from run_epoch import run_epoch, DataMode

/usr/local/lib/python3.11/dist-packages/transformers/utils/generic.py:441: FutureWarning: `torch.utils._pytree._register_pytree_node` is deprecated. Please use `torch.utils._pytree.register_pytree_node` instead.
  _torch_pytree._register_pytree_node(


In [ ]:
run_epoch(DataMode.BRAIN)

In [21]:
run_epoch(DataMode.BIG_BRAIN)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/file_download.py:949: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
100%|██████████| 582519/582519 [04:18<00:00, 2251.45it/s]
/workspace/efdl_hw/week02_fast_pipelines/homework/./task2/run_epoch.py:84: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler()
100%|██████████| 72815/72815 [25:59<00:00, 46.69it/s]


In [7]:
run_epoch(DataMode.ULTRA_BIG_BRAIN, k=1)

100%|██████████| 72968/72968 [16:46<00:00, 72.49it/s] 


In [8]:
run_epoch(DataMode.ULTRA_BIG_BRAIN, k=5)

100%|██████████| 72855/72855 [13:39<00:00, 88.89it/s] 


In [9]:
run_epoch(DataMode.ULTRA_BIG_BRAIN, k=10)

100%|██████████| 72845/72845 [13:54<00:00, 87.27it/s] 


In [10]:
run_epoch(DataMode.ULTRA_BIG_BRAIN, k=20)

100%|██████████| 72828/72828 [14:43<00:00, 82.43it/s] 


In [11]:
run_epoch(DataMode.ULTRA_BIG_BRAIN, k=50)

100%|██████████| 72822/72822 [15:21<00:00, 79.05it/s] 


In [12]:
run_epoch(DataMode.ULTRA_BIG_BRAIN, k=640)

100%|██████████| 72815/72815 [26:01<00:00, 46.64it/s]


In [13]:
run_epoch(DataMode.ULTRA_DUPER_BIG_BRAIN_NAIVE)

  0%|          | 0/181602 [00:00<?, ?it/s]/workspace/efdl_hw/week02_fast_pipelines/homework/./task2/dataset.py:119: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  'seq_mask': torch.tensor(seq_mask)
100%|██████████| 181602/181602 [34:53<00:00, 86.73it/s] 


In [14]:
run_epoch(DataMode.ULTRA_DUPER_BIG_BRAIN_FFD)

100%|██████████| 181375/181375 [32:23<00:00, 93.31it/s] 


In [15]:
run_epoch(DataMode.ULTRA_DUPER_BIG_BRAIN_OBFD)

100%|██████████| 181375/181375 [31:36<00:00, 95.65it/s] 


In [22]:
import pickle

brain_times = pickle.load(open('BRAIN.pkl', 'rb'))['times']
big_brain_times = pickle.load(open('BIG_BRAIN.pkl', 'rb'))['times']
ultra_big_brain_k1_times = pickle.load(open('ULTRA_BIG_BRAIN_k1.pkl', 'rb'))['times']
ultra_big_brain_k5_times = pickle.load(open('ULTRA_BIG_BRAIN_k5.pkl', 'rb'))['times']
ultra_big_brain_k10_times = pickle.load(open('ULTRA_BIG_BRAIN_k10.pkl', 'rb'))['times']
ultra_big_brain_k20_times = pickle.load(open('ULTRA_BIG_BRAIN_k20.pkl', 'rb'))['times']
ultra_big_brain_k50_times = pickle.load(open('ULTRA_BIG_BRAIN_k50.pkl', 'rb'))['times']
ultra_big_brain_k640_times = pickle.load(open('ULTRA_BIG_BRAIN_k640.pkl', 'rb'))['times']
ultra_duper_big_brain_naive_times = pickle.load(open('ULTRA_DUPER_BIG_BRAIN_NAIVE.pkl', 'rb'))['times']
ultra_duper_big_brain_ffd_times = pickle.load(open('ULTRA_DUPER_BIG_BRAIN_FFD.pkl', 'rb'))['times']
ultra_duper_big_brain_obfd_times = pickle.load(open('ULTRA_DUPER_BIG_BRAIN_OBFD.pkl', 'rb'))['times']

In [23]:
brain_times = brain_times[100:]
big_brain_times = big_brain_times[100:]
ultra_big_brain_k1_times = ultra_big_brain_k1_times[100:]
ultra_big_brain_k5_times = ultra_big_brain_k5_times[100:]
ultra_big_brain_k10_times = ultra_big_brain_k10_times[100:]
ultra_big_brain_k20_times = ultra_big_brain_k20_times[100:]
ultra_big_brain_k50_times = ultra_big_brain_k50_times[100:]
ultra_big_brain_k640_times = ultra_big_brain_k640_times[100:]
ultra_duper_big_brain_naive_times = ultra_duper_big_brain_naive_times[100:]
ultra_duper_big_brain_ffd_times = ultra_duper_big_brain_ffd_times[100:]
ultra_duper_big_brain_obfd_times = ultra_duper_big_brain_obfd_times[100:]

In [24]:
import pandas as pd
import numpy as np

algo = [
    'brain', 'big_brain', 'ultra_big_brain_k1', 'ultra_big_brain_k5', 'ultra_big_brain_k10', 'ultra_big_brain_k20', 'ultra_big_brain_k50', 'ultra_big_brain_k640', 'ultra_duper_big_brain_naive', 'ultra_duper_big_brain_ffd', 'ultra_duper_big_brain_obfd'
]
min_time = [
    min(brain_times), min(big_brain_times), min(ultra_big_brain_k1_times), min(ultra_big_brain_k5_times), min(ultra_big_brain_k10_times), min(ultra_big_brain_k20_times), min(ultra_big_brain_k50_times), min(ultra_big_brain_k640_times), min(ultra_duper_big_brain_naive_times), min(ultra_duper_big_brain_ffd_times), min(ultra_duper_big_brain_obfd_times)
]
max_time = [
    max(brain_times), max(big_brain_times), max(ultra_big_brain_k1_times), max(ultra_big_brain_k5_times), max(ultra_big_brain_k10_times), max(ultra_big_brain_k20_times), max(ultra_big_brain_k50_times), max(ultra_big_brain_k640_times), max(ultra_duper_big_brain_naive_times), max(ultra_duper_big_brain_ffd_times), max(ultra_duper_big_brain_obfd_times)
]
mean_time = [
    np.mean(brain_times), np.mean(big_brain_times), np.mean(ultra_big_brain_k1_times), np.mean(ultra_big_brain_k5_times), np.mean(ultra_big_brain_k10_times), np.mean(ultra_big_brain_k20_times), np.mean(ultra_big_brain_k50_times), np.mean(ultra_big_brain_k640_times), np.mean(ultra_duper_big_brain_naive_times), np.mean(ultra_duper_big_brain_ffd_times), np.mean(ultra_duper_big_brain_obfd_times)
]

median_time = [
    np.median(brain_times), np.median(big_brain_times), np.median(ultra_big_brain_k1_times), np.median(ultra_big_brain_k5_times), np.median(ultra_big_brain_k10_times), np.median(ultra_big_brain_k20_times), np.median(ultra_big_brain_k50_times), np.median(ultra_big_brain_k640_times), np.median(ultra_duper_big_brain_naive_times), np.median(ultra_duper_big_brain_ffd_times), np.median(ultra_duper_big_brain_obfd_times)
]

In [25]:
results = pd.DataFrame({
    'algo': algo,
    'min_time': min_time,
    'max_time': max_time,
    'mean_time': mean_time,
    'median_time': median_time
})

In [26]:
results

algo  min_time  max_time  mean_time  median_time
0                         brain  0.015556  0.099936   0.031205     0.030109
1                     big_brain  0.003977  0.080792   0.020854     0.013884
2            ultra_big_brain_k1  0.001867  0.076643   0.013027     0.007637
3            ultra_big_brain_k5  0.001877  0.077531   0.010724     0.004971
4           ultra_big_brain_k10  0.001913  0.081822   0.010928     0.005143
5           ultra_big_brain_k20  0.001962  0.078338   0.011535     0.005889
6           ultra_big_brain_k50  0.001973  0.086323   0.012088     0.006805
7          ultra_big_brain_k640  0.003428  0.081248   0.020830     0.013931
8   ultra_duper_big_brain_naive  0.002586  0.068005   0.005126     0.002791
9     ultra_duper_big_brain_ffd  0.002582  0.068806   0.005033     0.002756
10   ultra_duper_big_brain_obfd  0.002590  0.082180   0.005011     0.002735

Имплементация:
brain - падим втупую каждый семпл

big brain - там просто колатор который смотрит на максимум

ultra big brain - бьем на бины, дальше пушим в свой бин и в конце шафлим, далее передаем в семплер для каждого бина сколько батчей ему надо и шафлим индексы бина далее вытаскиваем их и получаем бин и из него уже индексы

ultra duper big brain naive - пушим семпл в конец последнего примера если не влезает то создаем еще один и туда засовываем вторую часть, еще надо сделать маску это блочная матрица где в диагональном блоке делаем обычную казуальную маску

ffd - тут за квадрат проходим в целом все как в описании задания

obfd:

делаем хешмапу которая из размера выдает все семплы в которых осталось ровно столько места

также делаем ДО на то сколько семплов есть с конкретным количеством места, так мы можем за лог отвечать на запрос какое минимальное место есть что хотябы заданое количество

далее идем по размерам от самого большого делаем запрос в ДО - он выдает лучший размер идем в хешмапу берем крайний индекс в листе чтобы легко удалить засовываем туда, обновляем то в какой лист в хешмапе он идет и обновляем два значения в ДО

если же нет подходящих мест создаем новый семпл кладем туда понимаем где в хешмапе он лежит и обновляем ДО


быстрее всего работают итерации ultra_duper_big_brain но это и понятно тк там у нас bs = 1, при этом если посмотреть сколько всего батчей нужно обработать (181375 / 72845 = 2.4898) то оказывается чуть дольше чем ultra_big_brain но все еще быстрее чем brain и big_brain

быстрее всего отрабатывает ultra_big_brain_k1 там не так много лишних батчей создается, но при этом каждый почти идеально пакуется

среднее за ultra_big_brain_k640 и big_brain одинаковое как и должно быть

также максимальные батчи у всех кроме big_brain схожи тк даже если упаковка оптимальная там все-равно может быть что куча больших текстов и батч максимальный 